### PR TITLE
feat: fast exact name search by default

### PR DESCRIPTION
**Pull request by lidel**
# tldr

This PR changes the default behavior of the `name` filter 

from **case-insensitive, partial match** 
to much faster **case-sensitive, exact (full)  match**

and adds `match` parameter that enables opt-in into alternative strategy when performance is not an object:

> Screenshot_2020-10-26 IPFS Pinning Service API

**PREVIEW:** https://ipfs.github.io/pinning-services-api-spec/#specUrl=https://raw.githubusercontent.com/ipfs/pinning-services-api-spec/fix/default-to-fast-name-match/ipfs-pinning-service.yaml

# Rationale for this change

- **Case-insensitive partial "Fuzzy" search (`ipartial`) makes a bad default. (performance-wise)** 
  Database indexes and b-trees do not solve the performance fully.
  Exact match is always faster and less expensive.

- **Fuzzy search may produce bugs**
  This PR introduces `match` query parameter, so developers who need fuzzy search  can still opt-in into slower, but more flexible text matching strategy by passing `match` parameter. 

- **We are moving away from slow defaults in IPFS ecosystem. Better to do this now, than later.**
  - A good example of [tyranny of small decisions](https://en.wikipedia.org/wiki/Tyranny_of_small_decisions) is `ipfs pin ls`, which is  abysmally slow when one has >1k pins. Nearly all pins in real life are recursive, and `ipfs pin ls --type=recursive` executes instantly, but is rarely used due to not being the default. 
  - Future `ipfs pin remote` and `ipfs pin local` APIs will be inspired by this spec and will have `name` attribute, however we don't want to  re-introduce bad choices into IPFS ecosystem. The default should be the fastest option available, and in  this case it is the exact `name` match.

# Implementation notes and migration plan

- This change does not change the wire format and does not impact ongoing MVP integration in go-ipfs and ipfs-webui (https://github.com/ipfs/ipfs-gui/issues/91, https://github.com/ipfs/go-ipfs/issues/7559), but we want to include it in this spec to ensure MVP does the right thing.

- Existing Pinning Services already implement the more difficult <del>`fuzzy`</del> `ipartial` strategy. Adding support for `match` filter and implementing much simpler `exact` strategy should be a small task, but will save everyone headache in the long run.

---- 

@aschmahmann @jacobheun @petar @gammazero @obo20 @andrew  @GregTheGreek @priom @jsign  @sanderpick @andrewxhill @ipfs/wg-pinning-services

### CODE REVIEW

File: ipfs-pinning-service.yaml [hunk 1]
```diff
@@ -461,6 +462,14 @@ components:
       example:
         status_details: "Queue position: 7 of 9" # PinStatus.info[status_details], when status=queued
 
+    TextMatchingStrategy:
+      description: Alternative text matching strategy
+      type: string
+      default: exact
+      enum:
+        - exact   # full match, case-sensitive (the implicit default)
+        - fuzzy   # partial match, case-insensitive
```
Comment (aschmahmann) [inline ipfs-pinning-service.yaml:471]: :bike: + :derelict_house: :
"fuzzy" I like that it's short, but not so happy that it's not quite precise in what it does. I'm fine with fuzzy, but putting out the call for better names before we decide we're just going with this one.

**Review by aschmahmann — APPROVED** [review]
I agree, this seems much more sensible from an ecosystem perspective.

File: ipfs-pinning-service.yaml (hunk 1 above)
Comment (lidel) [reply ipfs-pinning-service.yaml:471]: I am totally fine with renaming `fuzzy` to something better. 
Food for thought: `partial`, `relaxed`, `inexact`,  `loose`

Comment (jessicaschilling) [reply ipfs-pinning-service.yaml:471]: +1 for `partial` unless you specifically need to indicate it's case-insensitive too (if the latter, would suggest something like `full-sensitive` and `partial-insensitive` but that gets a little too bikesheddy, probably)

**Review by jessicaschilling — APPROVED** [review]

File: ipfs-pinning-service.yaml (hunk 1 above)
Comment (jacobheun) [reply ipfs-pinning-service.yaml:471]: +1 for `partial`

**Review by jacobheun — APPROVED** [review]

**Comment by achingbrain** [comment]
Do we need a `match` parameter at all?  Can we not just look at the passed `name` and if it contains a `*` character it's a wildcard match and if not it's exact?  Case sensitive/insensitive could be an option?

**Comment by lidel** [comment]
@achingbrain `*` could work, but we still need some means of making the search case-insensitive. 

If we replace `match` with support for `*` and optional `case=sensitive|insensitive`, then services need to implement 4 options instead of two (exact-casesensitive, exact-caseinsensitive, partial-casesensitive, partial-caseinsensitive).

File: ipfs-pinning-service.yaml (hunk 1 above)
Comment (lidel) [reply ipfs-pinning-service.yaml]: Renamed to `partial` in [`91299cc` (#66)](https://github.com/ipfs/pinning-services-api-spec/pull/66/commits/91299ccbdba9b1a6808d58ab828ecc2772b9efdf).

We need to figure out if thats enough, or do we want something more fancy like suggestion from https://github.com/ipfs/pinning-services-api-spec/pull/66#issuecomment-712134357

**Review by andrew — APPROVED** [review]

**Comment by andrew** [comment]
PR for implementation on the ruby pinning api server over here: https://github.com/ipfs-shipyard/rb-pinning-service-api/pull/5

**Comment by obo20** [comment]
This PR looks great to me. The changes will be helpful in keeping things performant by default while also allowing for some flexibility.
